File: 098-is-valid-BST/solution.go
/*
前序遍歷看js
中序遍歷看java
後序便利看golang
*/

package main

import (
	"gopractice/datastructure/Tree"
	"math"
)

type TreeNode = Tree.TreeNode

func isValidBST(root *TreeNode) bool {
	return dfs(root, math.MinInt, math.MaxInt)
}

func dfs(root *TreeNode, lower int, upper int) bool {
	if root == nil {
		return true
	}
	isLeftBST := dfs(root.Left, lower, root.Val.(int))
	isRightBST := dfs(root.Right, root.Val.(int), upper)
	if root.Val.(int) <= lower || root.Val.(int) >= upper {
		return false
	}
	return isLeftBST && isRightBST
}

File: 098-is-valid-BST/solution.py
from typing import Optional
from datastructure  import TreeNode

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def valid(root, lower, upper):
            if not root:
                return True
            is_left_bst = valid(root.left, lower, root.val)
            is_right_bst = valid(root.right, root.val, upper)
            if root.val <= lower or root.val >= upper:
                return False
            return is_left_bst and is_right_bst
        return valid(root, float('-inf'), float('inf'))